nova_vm/src/ecmascript/scripts_and_modules/eval_source.rs

```diff
@@ -0,0 +1,113 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! EvalSource is a Nova-engine specific concept to capture and keep any
+//! `eval(source)` source strings alive after the eval call for the case where
+//! that the eval call defines functions. Those functions will refer to the
+//! EvalSource for their function source text.
+
+use std::{fmt::Debug, ops::Index, ptr::NonNull};
+
+use oxc_allocator::Allocator;
+
+use crate::{
+    ecmascript::{execution::Agent, types::HeapString},
+    heap::{
+        indexes::BaseIndex, CompactionLists, CreateHeapData, Heap, HeapMarkAndSweep, WorkQueues,
+    },
+};
+
+type EvalSourceIndex = BaseIndex<EvalSourceHeapData>;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub(crate) struct EvalSource(EvalSourceIndex);
```

Comment:
**question:** Is the plan to eventually extend `EvalSource` to scripts as well, so scripts can be GC'd if they're not referenced by anything? In that case, we should probably name this `ScriptSource` or something like that.

Reply:
Scripts already can be GC'd if they're not referenced if no functions reference them.

Scripts have quite a bit more stuff going on with them. I _think_ the way forward with this sort of setup for scripts would be to just add these equivalent fields to ScriptHeapData.

So in effect: This would stay as eval-source only.

Reply:
In test262 there's `$262.evalScript()`, which IIUC should create a script record from a JS string source. In HTML you can also create a script from a JS string. How does that play into all of this?

Reply:
Oh, ugh, oof.

I mean, I guess it mostly wouldn't be any different from calling `parse_script` and all that stuff, except that we'd move creating the `Allocator` inside that method entirely and keep it in the `ScriptHeapData` similarly to how we keep it for eval'd data in `EvalSource`. The `parse_script` would also probably change from taking a `Box<str>` to taking a `std::string::String` or an alternative API for taking `&'static str`, or maybe just going broke and taking a `nova_vm::String` directly? And then we'd keep a reference to that inside the `ScriptHeapData`, replacing the current `Box<str>` there.

Reply:
Yes, that's what I'm talking about in the latter half. The allocator would move inside the heap similarly to how it does here for eval calls. It would also take a JS String as the parameter.

If you want, I can use this PR to change the script running API as well so that it handles the Script Allocator creation inside of the heap.

Reply:
That doesn't need to be in this PR.

Reply:
Alright, I'll open a new PR for that.